### DISCUSSION TITLE
Is there a way to build a URL or request based on route information?

### DISCUSSION THREAD

**Discussion by tmcw** [fastify/fastify#5565] [Q&A]
I'm starting to implement some [json api-style links](https://jsonapi.org/format/#document-links) in my API, like in their example:

```json
"links": {
  "self": "http://example.com/articles/1/relationships/comments"
}
```

So each item in a list of things returned from the API might have this links object that provides a URL to where you'd run a GET request to get just that object individually. I'd love to have a way to implement this in fastify without having to explicitly redefine the URL paths of other routes - for example a naive implementation might be:

```ts
reply.send({
  ...object
  "links": {
    "self": `http://example.com/articles/${id}/relationships/comments`
  }
});
```

But this would mean that I if I typo `articles` in that URL, or change the route definition, that URL would no longer be valid. There are some examples of ways to solve for this, like:

- Rails's `url_for`: https://api.rubyonrails.org/v7.1.3.4/classes/ActionDispatch/Routing/UrlFor.html
- Remix's typed-routes: https://github.com/yesmeck/remix-routes

An API for this might look like:

```ts
reply.send({
  ...object
  "links": {
    "self": fastify.urlFor('comments', id)
  }
});
```

In which `comments` is a the route `.config.id` parameter. Maybe this thing already exists?

**Comment by WhereJuly** [comment]
@tmcw 
You could refer to [Request](https://fastify.dev/docs/v4.28.x/Reference/Request/) docs and explore the request content. 

May check `request.routeConfig` . E.g. `request.routeConfig.url` contains the route URL like `/articles/:id/relationships/comments`.

You could regex replace the dynamic values (e.g. ID) to construct the actual URL (and take host from `request.hostname`).
